operator_rolling_upgrades.md

```diff
@@ -0,0 +1,54 @@
+# Operator Rolling Upgrades
+Optionally allow a Infinispan CR cluster to be upgraded with zero downtime and no loss of data.
+
+## Spec
+Add a field to the CR spec that determines whether an upgrade should be `rolling` or `gracefulShutdown`. The latter is the current approach that does not allow data migration.
```

Comment:
If the user picks `rolling`, will it roll via **procedure** described below for every release or would it be achievable for some of them to be done through in-place (eg. micro from `11.0.1` to `11.0.2`, security fixes to underlying images that won't cause version change at all)?

If the user decides to begin update process from eg 11.0.0 to 12.0.0 and there are 3 more releases to 11. The whole process will be repeated 5 times `11.0.0` -> `11.0.1` -> `11.0.2` -> `11.0.3` -> `12.0.0` which could be quite time (and resource) consuming. Question is, if we managed to do at least some of the `11.0.0` -> `11.0.3` inplace, would it save some time and made the process more seamless?

Another possibility of improvement here is start using `skip` field in CSVs to allow direct update from `11.0.0` to `12.0.0`

Reply:
> If the user picks `rolling`, will it roll via **procedure** described below for every release or would it be achievable for some of them to be done through in-place (eg. micro from `11.0.1` to `11.0.2`, security fixes to underlying images that won't cause version change at all)?

It's an interesting idea for changes to the layers below our image. However, I don't think it would be possible for the patch upgrades to be in-place without us having all of the Infinispan core capabilities and tests that are required for the full in-place rolling upgrade design.



> If the user decides to begin update process from eg 11.0.0 to 12.0.0 and there are 3 more releases to 11. The whole process will be repeated 5 times `11.0.0` -> `11.0.1` -> `11.0.2` -> `11.0.3` -> `12.0.0` which could be quite time (and resource) consuming. Question is, if we managed to do at least some of the `11.0.0` -> `11.0.3` inplace, would it save some time and made the process more seamless?

It would definitely save time for the user, but I fear that this would also add considerable complexity to what is designed to be a quick solution to a larger problem, i.e. solving in-place rolling upgrades.



> Another possibility of improvement here is start using `skip` field in CSVs to allow direct update from `11.0.0` to `12.0.0`

Sounds good to me, but I'm lacking knowledge about OLM to know how this works.

Reply:
> > If the user picks `rolling`, will it roll via **procedure** described below for every release or would it be achievable for some of them to be done through in-place (eg. micro from `11.0.1` to `11.0.2`, security fixes to underlying images that won't cause version change at all)?
> 
> It's an interesting idea for changes to the layers below our image. However, I don't think it would be possible for the patch upgrades to be in-place without us having all of the Infinispan core capabilities and tests that are required for the full in-place rolling upgrade design.

We can pretest this combination and store all possible combinations as a graph for upgrade options definition

Reply:
The problem is that infinispan's core is lacking some of the capabilities required to correctly handle in-place upgrades. This is a long term goal, but is beyond the scope of this design document.

Reply:
Right, I don't think it makes sense to have some updates on the channel as an in-place and some via RollingUpgrade. It would unnecessarily complicate the code to recognize which is possible and since  in-place are driven by SS it's spec would change all the time.